src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendReceive.cs

```diff
@@ -0,0 +1,555 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Sockets.Tests
+{
+    public class SendReceive
+    {
+        private readonly ITestOutputHelper _log;
+
+        public SendReceive(ITestOutputHelper output)
+        {
+            _log = output;
+        }
+
+        private static void SendToRecvFrom_Datagram_UDP(IPAddress leftAddress, IPAddress rightAddress)
+        {
+            const int DatagramSize = 256;
+            const int DatagramsToSend = 256;
+
+            var left = new Socket(leftAddress.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
+            left.BindToAnonymousPort(leftAddress);
+
+            var right = new Socket(rightAddress.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
+            right.BindToAnonymousPort(rightAddress);
+
+            var leftEndpoint = (IPEndPoint)left.LocalEndPoint;
+            var rightEndpoint = (IPEndPoint)right.LocalEndPoint;
+
+            var receiverAck = new ManualResetEventSlim();
+            var senderAck = new ManualResetEventSlim();
+
+            var receivedChecksums = new uint[DatagramsToSend];
+            var leftThread = new Thread(() =>
+            {
+                using (left)
+                {
+                    EndPoint remote = leftEndpoint.Create(leftEndpoint.Serialize());
+                    var recvBuffer = new byte[DatagramSize];
+                    for (int i = 0; i < DatagramsToSend; i++)
+                    {
+                        int received = left.ReceiveFrom(recvBuffer, SocketFlags.None, ref remote);
+                        Assert.Equal(DatagramSize, received);
+                        Assert.Equal(rightEndpoint, remote);
+                        receivedChecksums[(int)recvBuffer[0]] = Fletcher32.Checksum(recvBuffer, 0, received);
+
+                        receiverAck.Set();
+                        senderAck.Wait();
```

Comment:
I recommend setting some reasonable time limits to all `Wait()` and `Join()` statements. 
(Coincidentally, we had a failure in our internal tests caused by these limits missing.)